### PR TITLE
Glue kernel-space backtrace with user-space backtrace.

### PR DESCRIPTION
**Pull request by mzr**
Providing unwind debug info using .cfi_* directives.
https://sourceware.org/binutils/docs/as/CFI-directives.html#CFI-directives
#553

### CODE REVIEW

File: include/mips/exc.h [hunk 1]
```diff
@@ -7,7 +7,18 @@
...
-#define SAVE_REG(reg, offset, base) sw reg, (EXC_##offset)(base)
+#define _ADD_CFI_NO_CFI(reg, offset)
+
+#define _ADD_CFI_(reg, offset) .cfi_rel_offset reg, offset
+
+/* Resolves to _ADD_CFI_(reg, offset) when no variadic arguments given.
+ * Resolves to _ADD_CFI_NO_CFI when NO_CFI given as one, and only one
+ * variadic argument.
+ * Passing multiple variadic arguments to this macro is consiered an error.
+ */
```
Comment (jpszczolowski) [inline include/mips/exc.h]: `consiered` typo

File: include/Makefile [hunk 2]
```diff
@@ -6,8 +6,7 @@ HEADERS = $(wildcard *.h) $(foreach dir, machine sys, $(wildcard $(dir)/*.h))
 INSTALL-FILES = $(addprefix $(SYSROOT)/usr/include/, $(HEADERS))
 
 SOURCES_H = $(foreach dir, . dev mips sys, $(wildcard $(dir)/*.h))
-FORMAT-EXCLUDE = mips/asm.h mips/m32c0.h \
-                 sys/syscall.h sys/syscallargs.h
+FORMAT-EXCLUDE = mips/m32c0.h sys/syscall.h sys/syscallargs.h
```
Comment (jpszczolowski) [inline include/Makefile]: Maybe you could move reformatting of `include/mips/asm.h` to a different PR?

File: sys/mips/exc.S [hunk 3]
```diff
@@ -21,36 +21,57 @@
...
 # [$k0] must be set to value of C0_STATUS
+//NESTED(exc_enter, 0, sp)
 exc_enter:
         mfc0    k0, C0_STATUS           
         andi    k0, SR_KSU_MASK         # Did exception occur in kernel mode?
         beqz    k0, kern_exc_enter
         nop
+        j user_exc_enter
+        nop
+
```
Comment (jpszczolowski) [inline sys/mips/exc.S]: Why do we jump to `user_exc_enter` here instead of falling-through?

File: sys/mips/exc.S [hunk 4]
```diff
@@ -21,36 +21,57 @@
...
-        lw      k0, PCPU_KSP(k0)
 
-        SAVE_CPU_CTX(sp, k0)
+        # sp contains kernel stack pointer
+        lw      sp, PCPU_KSP(k0)
+
+        .cfi_def_cfa sp, 0
+
+        // save user sp
+        sw k1, 104(sp)
```
Comment (jpszczolowski) [inline sys/mips/exc.S]: Inconsistent comment style, maybe use `#` instead of `//`

File: sys/mips/exc.S [hunk 5]
```diff
@@ -21,36 +21,57 @@
         .local skip_fpu_save
         .local skip_fpu_restore
 
+        .cfi_sections .debug_frame, .eh_frame
+
 # [$k0] must be set to value of C0_STATUS
+//NESTED(exc_enter, 0, sp)
 exc_enter:
```
Comment (jpszczolowski) [inline sys/mips/exc.S]: What about this `//NESTED(exc_enter, 0, sp)`?

File: include/Makefile (hunk 2 above)
Comment (cahirwpz) [reply include/Makefile]: Did that on my own. Please clean up `mips/asm.h` header file.

File: include/mips/asm.h [hunk 6]
```diff
@@ -119,33 +87,36 @@
...
+  .cfi_startproc
 
 /*
- * XLEAF
- *	declare alternate entry to leaf routine
- */
+ *  * XLEAF
+ *   *  declare alternate entry to leaf routine
+ *    */
 #define XLEAF(x)                                                               \
```
Comment (jpszczolowski) [inline include/mips/asm.h]: Unnecessary asterisks.

File: include/mips/asm.h [hunk 7]
```diff
@@ -119,33 +87,36 @@
...
+  .cfi_startproc
 
 /*
- * LEAF_NOPROFILE
- *	No profilable leaf routine.
- */
+ *  * LEAF_NOPROFILE
+ *   *  No profilable leaf routine.
+ *    */
 #define LEAF_NOPROFILE(x)                                                      \
```
Comment (jpszczolowski) [inline include/mips/asm.h]: Unnecessary asterisks.

File: include/mips/asm.h [hunk 8]
```diff
@@ -172,7 +176,8 @@
   .globl _C_LABEL(x);                                                          \
   .ent _C_LABEL(x), 0;                                                         \
   _C_LABEL(x) :;                                                               \
-  .frame sp, fsize, retpc
+  .frame sp, fsize, retpc;                                                     \
+  .cfi_startpoic
```
Comment (cahirwpz) [inline include/mips/asm.h]: Typo!

File: include/mips/exc.h [hunk 9]
```diff
@@ -7,7 +7,18 @@
...
+#define _ADD_CFI_NO_CFI(reg, offset)
+
+#define _ADD_CFI_(reg, offset) .cfi_rel_offset reg, offset
+
+/* Resolves to _ADD_CFI_(reg, offset) when no variadic arguments given.
+ * Resolves to _ADD_CFI_NO_CFI when NO_CFI given as one, and only one
+ * variadic argument.
+ * Passing multiple variadic arguments to this macro is considered an error.
+ */
+#define SAVE_REG(reg, offset, base, ...)                                       \
```
Comment (cahirwpz) [inline include/mips/exc.h]: Could you be more direct and rename it to `SAVE_REG_CFI`, restore original `SAVE_REG` and remove `_ADD_CFI_*` clutter?

File: include/mips/exc.h [hunk 10]
```diff
@@ -56,7 +72,10 @@
   SAVE_REG(t0, SR, reg);                                                       \
   SAVE_REG(t1, CAUSE, reg);                                                    \
   SAVE_REG(t2, BADVADDR, reg);                                                 \
-  SAVE_REG(t3, PC, reg)
+  /* Saving value of user-space PC just before syscall. */                     \
+  SAVE_REG(t3, PC, reg, NO_CFI);                                               \
+  .cfi_return_column t3;                                                       \
```
Comment (cahirwpz) [inline include/mips/exc.h]: What does `.cfi_return_column` do ?

File: lib/libc/sys/syscalls.S [hunk 11]
```diff
@@ -1,5 +1,7 @@
 #include <machine/syscall.h>
 
+.cfi_sections .debug_frame, .eh_frame
```
Comment (cahirwpz) [inline lib/libc/sys/syscalls.S]: I removed those entirely, rebuild repo from scratch and kernel-user transition in backtrace still works. Are those really needed? What they were supposed to do then?

File: sys/mips/ebase.S [hunk 12]
```diff
@@ -15,6 +15,8 @@
 
         .section .ebase, "ax"
 
+        .cfi_sections .debug_frame, .eh_frame
```
Comment (cahirwpz) [inline sys/mips/ebase.S]: ditto

File: sys/mips/exc.S [hunk 13]
```diff
@@ -21,36 +21,48 @@
         .local skip_fpu_save
         .local skip_fpu_restore
 
+        .cfi_sections .debug_frame, .eh_frame
```
Comment (cahirwpz) [inline sys/mips/exc.S]: ditto

File: sys/mips/exc.S [hunk 14]
```diff
@@ -21,36 +21,48 @@
...
         beqz    k0, kern_exc_enter
         nop
 
+
 # IMPORTANT: While exception level is active the code may only access kernel
 # stack and local pcpu structure! FTTB These are accessed through kseg0.
 # It is UNSAFE to use k0/k1 registers when TLB miss handler may be triggered!
 
 NESTED(user_exc_enter, USER_EXC_FRAME_SIZE, ra)
+        .cfi_signal_frame
```
Comment (cahirwpz) [inline sys/mips/exc.S]: This one is really required. Without it backtrace stops at first stack frame in user space. Good job!

File: sys/mips/exc.S [hunk 15]
```diff
@@ -21,36 +21,48 @@
...
+
+
+        # Move user sp to temporary register and replace it with kernel sp.
+        move    k1, sp
+        .cfi_register sp, k1
```
Comment (cahirwpz) [inline sys/mips/exc.S]: Could you add comments to `.cfi_register` and `.cfi_def_cfa`. I reckon nobody will remember what the directives do in 6 months.

File: sys/mips/exc.S [hunk 16]
```diff
@@ -21,36 +21,48 @@
...
+        .cfi_sections .debug_frame, .eh_frame
+
 # [$k0] must be set to value of C0_STATUS
 exc_enter:
         mfc0    k0, C0_STATUS           
         andi    k0, SR_KSU_MASK         # Did exception occur in kernel mode?
         beqz    k0, kern_exc_enter
         nop
 
+
```
Comment (cahirwpz) [inline sys/mips/exc.S]: Could you remove extranous empty lines in this file?

File: sys/mips/exc.S [hunk 17]
```diff
@@ -21,36 +21,48 @@
...
+        # Exception frame will be saved on kernel stack. 
         # Fetch exception frame pointer where context will be saved.
         LOAD_PCPU_KSEG0(k0)
-        lw      k0, PCPU_KSP(k0)
 
-        SAVE_CPU_CTX(sp, k0)
+        # sp contains kernel stack pointer
+        lw      sp, PCPU_KSP(k0)
+
+        .cfi_def_cfa sp, 0
```
Comment (cahirwpz) [inline sys/mips/exc.S]: If a directive is associated with given instruction do not insert newline between them.

File: build/flags.mk [hunk 18]
```diff
@@ -1,4 +1,4 @@
 ASFLAGS  +=
 WFLAGS   += -Wall -Wextra -Wno-unused-parameter -Wstrict-prototypes -Werror
-CFLAGS   += -std=gnu11 -Og -ggdb3 $(WFLAGS)
+CFLAGS   += -std=gnu11 -Og -ggdb3 $(WFLAGS) -fno-asynchronous-unwind-tables
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply build/flags.mk]: Seems that everything works when it's removed.

File: lib/libc/sys/syscalls.S (hunk 11 above)
Comment (mzr) [reply lib/libc/sys/syscalls.S]: According to [doc](https://sourceware.org/binutils/docs/as/CFI-directives.html#CFI-directives):

> `.cfi_sections` may be used to specify whether CFI directives should emit `.eh_frame` section and/or `.debug_frame` section. If section_list is `.eh_frame`, `.eh_frame` is emitted, if section_list is `.debug_frame`, `.debug_frame` is emitted. To emit both use `.eh_frame`, `.debug_frame`. 

But:
> The default if this directive is not used is .cfi_sections .eh_frame.

So we should be fine without these.

File: include/mips/exc.h (hunk 10 above)
Comment (mzr) [reply include/mips/exc.h]: According to doc: 
> Change return column register, i.e. the return address is either directly in register or can be accessed by rules for register.

I only use it to later insert `.cfi_rel_offset t3, EXC_PC`. To my best knowledge this is the only way to specify return address to instruction just before entering kernel-space.

Comment (mzr) [reply include/mips/exc.h]: Without this we miss last frame before entering kernel-space. I.e. `write` which calls syscall.

**Pull request merged**